### DISCUSSION TITLE
Linux start-up extremely slow

### DISCUSSION THREAD

**Discussion by honza** [ghostty-org/ghostty#3752] [Q&A] [answered]
I'm on NixOS. I cloned the repo and ran:

```
$ nix-shell
$ zig build
$ ./zig-out/bin/ghostty
```

It takes over 30s to see the GUI window. It appears the time is spent in GTK. The same happens in a release build.

Once running, the program is fast.

The same happens when I use the flake.

I'm using Wayland (sway) without full GNOME.

Any hints on what I could try?

**Comment by EricDriussi** [comment]
I'm seeing the same behavior on a completely different system: arch using i3 with MATE on xorg.
I'm using an nvidia graphics card in case that helps.

Running from another terminal gives me a couple of warnings/errors that don't really seem major but I'll drop them in here in case it's useful:

```
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
warning(gtk): unimplemented action=apprt.action.Action.Key.size_limit
info(io_exec): started subcommand path=/bin/sh pid=152371
info(io_exec): subcommand cgroup=-
error(gtk): unable to get current color scheme: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such interface “org.freedesktop.portal.Settings” on object at path /org/freedesktop/portal/desktop
```

I can drop the whole output if needed.

Maybe it has to do with GTK, I'm pretty sure both GNOME and  MATE use that toolkit.

**Reply by rhodes-b** [reply]
There might be some slowdowns on systems that dont use DBUS but it shouldnt be extreme (like multiple seconds worth)

**Comment by 00-kat** [comment]
First up:

> `zig build`

Use a release build (add `-Doptimize=ReleaseFast`).  Debug builds have some very extravagant checks which slow it down *considerably*.  (I remember it took around a minute for a window to spawn when I tried a debug build, as opposed to under a second with release builds.)

Secondly, you could try [`gtk-single-instance`](https://ghostty.org/docs/help/gtk-single-instance) to aid with the time it takes for windows after the first one to open.

**Comment by honza** [comment]
Setting `gtk-single-instance = true` doesn't seem to have any effect.

**Reply by SuperDuperDeou** [reply]
Can you post the output of `ghostty +version`? If the build mode isn't ReleaseFast, it being slow is expected.

**Reply by rhodes-b** [reply]
To be clear after the first one `gtk-single-instance = true` doesnt help? (this should only effect after the first time you launch ghostty)

can you get some kind of trace with `flamegraph` or another trace tool you are familiar with

**Reply by honza** [reply]
I'd be happy to provide any details but unfortunately, I can't say I'm familiar with tools like that. Would you mind giving me a rough cli incantation?

**Comment by honza** [comment]
As I said in the OP, I tried an optimized build and the result is the same.

**Comment by honza** [comment]
```
info: ghostty version=1.0.1-main+d5481760
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_US.UTF-8
info(gtk): GTK version build=4.16.3 runtime=4.16.3
info: reading configuration file path=/home/honza/.config/ghostty/config
info(config): default shell source=env value=/run/current-system/sw/bin/fish
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2
info(gtk_systemd_cgroup): transient scope created cgroup=/user.slice/user-1000.slice/user@1000.service/app.slice/app-ghostty-transient-2403554.scope
info(gtk): cgroup isolation enabled base=/user.slice/user-1000.slice/user@1000.service/app.slice/app-ghostty-transient-2403554.scope
error(gtk): unable to get current color scheme: Timeout was reached
EGL API version 1.5 found
 - Vendor: Mesa Project
 - Version: 1.5
 - Device: /dev/dri/renderD128, 226 128
 - Client APIs: OpenGL OpenGL_ES
```

**Comment by MunsMan** [comment]
I noticed the same issue on NixOS with HyprLand. I'm using the nix flake provided in the documentation.
Spawning a new application instance takes around 20–30 seconds.

**Comment by honza** [comment]
It looks like this is related to dbus.  There is a [hint in the sway wiki](https://github.com/swaywm/sway/wiki#gtk-applications-take-20-seconds-to-start).

**Reply by EricDriussi** [reply]
Maybe, but I'm seeing the same issue on xorg using the official arch package

**Comment by honza** [comment]
OK, it was dbus for me.  Sway wasn't properly configured for GTK/GNOME. Ghostty now starts instantly.

**Comment by pluiedev** [comment, marked as answer]
For the record, this is fixed in 1.1 (#5064), but you should install xdg-desktop-portal-gtk to address the root cause.

**Comment by nonetrix** [comment]
Startup speed could still be improved in my opinion, and I'm 99% sure I have everything configured correctly. It's about one second for me on an SSD, which might not sound like a lot, but it feels much slower than other terminals which are so fast I can't measure it easily. Bandaid solution would be to add a setting to run as a daemon since I find `gtk-single-instance` bridges the gap, but I have to have at least one terminal open for that to work

**Reply by 00-kat** [reply]
I remember there being discussion about a SystemD service to run `ghostty --gtk-single-instance=true --quit-after-last-window-closed=false --initial-window=false`, but I can't remember *where* that was being discussed though.

**Reply by rhodes-b** [reply]
https://discord.com/channels/1005603569187160125/1322831013746970624

**Reply by nonetrix** [reply]
Ah I guess I can just put that in my Hyprland config for now, but be nice if root of the issue could be fixed

**Reply by 00-kat** [reply]
FYI in case that doesn't make it any faster for you try without `--initial-window=false`; something like
```ini
exec-once = [workspace special:hallofshame silent] ghostty --gtk-single-instance=true --quit-after-last-window-closed=false
```
